flang/lib/Optimizer/Dialect/FIROps.cpp

```diff
@@ -3677,10 +3677,14 @@ fir::parseSelector(mlir::OpAsmParser &parser, mlir::OperationState &result,
   return mlir::success();
 }
 
-mlir::func::FuncOp
-fir::createFuncOp(mlir::Location loc, mlir::ModuleOp module,
-                  llvm::StringRef name, mlir::FunctionType type,
-                  llvm::ArrayRef<mlir::NamedAttribute> attrs) {
+mlir::func::FuncOp fir::createFuncOp(mlir::Location loc, mlir::ModuleOp module,
+                                     llvm::StringRef name,
+                                     mlir::FunctionType type,
+                                     llvm::ArrayRef<mlir::NamedAttribute> attrs,
+                                     const mlir::SymbolTable *symbolTable) {
+  if (symbolTable)
+    if (auto f = symbolTable->lookup<mlir::func::FuncOp>(name))
+      return f;
```

Comment:
Is it possible to make sure that the `symbolTable` is kept in sync with the module?  E.g. can we put assertions here and other functions where we lookup symbols to verify that `symbolTable->lookup(name) == module.lookupSymbol(name)`.  Whenever the assertion fails this will indicate that there is code that adds/removes the symbol to/form the module and does not update the `symbolTable` - then we can track it down and fix.

Reply:
It is OK from a correctness point of view if symbols are not added to the symbolTable (when not found there, the code fallbacks to looking at the module), but you are right that if the symbols are deleted/replaced, that would be an issue. I added the suggested assert in the lookups (under EXPENSIVE_CHECKS, since doing the check kills any compilation time improvement of using the map).